core/src/main/scala/neotypes/Transaction.scala

```diff
@@ -14,105 +15,140 @@ import org.neo4j.driver.exceptions.NoSuchRecordException
-                                       (implicit F: Async[F], resultMapper: ResultMapper[T]): F[C] =
-    F.async { cb =>
-      transaction
-        .runAsync(query, QueryParam.toJavaMap(params))
-        .thenCompose(_.listAsync())
-        .accept(cb) { records =>
-          traverseFun(records.asScala.iterator) { record =>
+    override final def list[T](query: String, params: Map[String,QueryParam])
+                              (implicit resultMapper: ResultMapper[T]): F[List[T]] =
+      collectAsImpl(transaction, query, params)(traverseAsList[Record, T])
+
+    override final def map[K, V](query: String, params: Map[String,QueryParam])
+                                (implicit resultMapper: ResultMapper[(K, V)]): F[Map[K,V]] =
+      collectAsImpl(transaction, query, params)(traverseAsMap[Record, K, V])
+
+    override final def set[T](query: String, params: Map[String,QueryParam])
+                             (implicit resultMapper: ResultMapper[T]): F[Set[T]] =
+      collectAsImpl(transaction, query, params)(traverseAsSet[Record, T])
+
+    override final def vector[T](query: String, params: Map[String,QueryParam])
+                                (implicit resultMapper: ResultMapper[T]): F[Vector[T]] =
+      collectAsImpl(transaction, query, params)(traverseAsVector[Record, T])
+
+    override final def single[T](query: String, params: Map[String,QueryParam])
+                                (implicit resultMapper: ResultMapper[T]): F[T] =
+      F.async { cb =>
+        transaction
+          .runAsync(query, QueryParam.toJavaMap(params))
+          .thenCompose(_.singleAsync())
+          .acceptExceptionally(cb) { record =>
             resultMapper.to(recordToList(record), None)
+          } {
+            case _: NoSuchRecordException => resultMapper.to(List.empty, None)
+          }
+      }
+
+    private def nextAsyncToF[T](cs: CompletionStage[Record])
+                               (implicit resultMapper: ResultMapper[T]): F[Option[T]] =
+      F.async { cb =>
+        cs.accept(cb) { r =>
+          Option(r) match {
+            case None =>
+              Right(None)
+
+            case Some(record) =>
+              resultMapper.to(recordToList(record), None).map(r => Option(r))
           }
         }
-    }
+      }
+
+    override final def stream[T, S[_]](query: String, params: Map[String,QueryParam])
+                                      (implicit S: Stream.Aux[S,F], resultMapper: ResultMapper[T]): S[T] =
+      S.fToS(
+        F.async { cb =>
+          transaction
+            .runAsync(query, QueryParam.toJavaMap(params))
+            .accept(cb) { statementResultCursor =>
+              Right(S.init(() => nextAsyncToF(statementResultCursor.nextAsync())))
+            }
+        }
+      )
+
+    override final def commit: F[Unit] =
+      F.async[Unit] { cb =>
+        transaction.commitAsync().acceptVoid(cb)
+      } flatMap { _=>
```

Comment:
If `transaction.commitAsync().acceptVoid(cb)` fails, we still want to release the lock, don't we?

Reply:
You are goddamn right!
We need to use `guarantee` here.